### ISSUE TITLE
Missing scala.scalajs.runtime.RuntimeString$.toLowerCase(java.lang.String,java.util.Locale)

### ISSUE THREAD

**Issue by mkotsbak** [cquiroz/scala-java-locales#50]
I got to this when trying org.threeten.bp.format.DateTimeFormatter$.ofPattern (release "2.0.0-M1")

Full relevant part of stacktrace:

```
[error] Referring to non-existent method scala.scalajs.runtime.RuntimeString$.toLowerCase(java.lang.String,java.util.Locale)java.lang.String
[error]   called from org.threeten.bp.format.DateTimeFormatterBuilder$ZoneIdPrinterParser$SubstringTree.add(java.lang.String)scala.Unit
[error]   called from org.threeten.bp.format.DateTimeFormatterBuilder$ZoneIdPrinterParser$.org$threeten$bp$format$DateTimeFormatterBuilder$ZoneIdPrinterParser$$prepareParser(java.util.Set)org.threeten.bp.format.DateTimeFormatterBuilder$ZoneIdPrinterParser$SubstringTree
[error]   called from org.threeten.bp.format.DateTimeFormatterBuilder$ZoneIdPrinterParser.parse(org.threeten.bp.format.DateTimeParseContext,java.lang.CharSequence,scala.Int)scala.Int
[error]   called from org.threeten.bp.format.DateTimeFormatterBuilder$CompositePrinterParser$$anonfun$parse$1.apply$mcV$sp()scala.Unit
[error]   called from scala.util.control.Breaks.breakable(scala.Function0)scala.Unit
[error]   called from org.threeten.bp.format.DateTimeFormatterBuilder.parsePattern(java.lang.String)scala.Unit
[error]   called from org.threeten.bp.format.DateTimeFormatterBuilder.appendPattern(java.lang.String)org.threeten.bp.format.DateTimeFormatterBuilder
[error]   called from org.threeten.bp.format.DateTimeFormatter$.ofPattern(java.lang.String)org.threeten.bp.format.DateTimeFormatter
```

**Comment by mkotsbak** [comment]
Here is Apache Harmony implementation: http://www.docjar.com/html/api/java/lang/String$ConsolePrintStream.java.html

It seems like there are very few exceptions to just using normal tolowerCase that is in SJS implemented natively.

**Comment by mkotsbak** [comment]
There is actually a similar function in JS, but it uses the browser's Locale: http://www.w3schools.com/jsref/jsref_tolocalelowercase.asp

But maybe that one is better than nothing for now?

**Comment by cquiroz** [comment]
We have reviewed this case and an issue is that it should be part of the core String class, we could perhaps add it using an implicit class but then you'd need to import it

The browser's `toLocaleLowerCase` call is of no use in this case since it doesn't let you indicate the locale. I guess we should try to implement it regardless and find a way to make core call it

**Comment by mkotsbak** [comment]
I guess the main SJS project could could add the implementation once it is complete and stable. Until then an import probably will be needed for code using it.

**Comment by soc** [comment]
https://github.com/soc/scala-java-time/pull/38

**Comment by soc** [comment]
- Should we close this?
- If yes, should we create a new ticket in scala-js/scala-js?

**Comment by mkotsbak** [comment]
Let it be open, the general case is not solved, and it is closely related to other Locale stuff.

**Comment by cquiroz** [comment]
@sjrd Do you think there is a way to achieve this without integrating `scala-java-locales` into `scala.js`? Could for example `toLowerCase(String, Locale)` use `scala-java-locales` if present and fail/ignore otherwise

If not possible I think we should close this issue.

**Comment by sjrd** [comment]
Is there a public API (in the JDK API) of `java.util.Locale` or one the associated classes that `String.toLowerCase` can delegate to? Like a `(loc: Locale).toLowerCase(str: String)` that would be equivalent to `(str: String).toLowerCase(loc)`? If yes we could just call that method from core `String.toLowerCase(Locale)`; that method won't link, unless it is provided by a third party such as `scala-java-locales`.

Otherwise, I don't think there's any solution that would allow scala-java-locales to provide the implementation. The core would have to provide it. We can consider it if it is small enough and if it is an isolated case (if there are others like that, this won't scale). But then we would not be able to *test* it in the core (since the core cannot depend on scala-java-locales to provide `Locale` itself) so we would have an untestable implementation in the core. That's not nice, but if there is no other choice ... It would first have to be implemented and tested as a separate method in a test project (possibly in the tests of scala-java-locales itself) with the same level of excellence that is required for Core stuff, and reviewed by us (the core team); then we could take the implementation and leave the tests behind. Ideally scala-java-locales should keep the tests around and rewire them to `String.toLowerCase` once it depends on a version of the core that provides it.

**Comment by cquiroz** [comment]
I'm afraid there is no such thing. Though it is not explained on the docs I think the jvm just does it internally

The code is not that small especially as I parse the rules provided by CLDR

I'll just close this as won't fix
